src/libstd/fs.rs

```diff
@@ -0,0 +1,1474 @@
+/// it was opened with. Files also implement `Seek` to alter the logical cursor
+/// that the file contains internally.
+///
+/// # Example
+///
+/// ```no_run
+/// use std::io::prelude::*;
+/// use std::fs::File;
+///
+/// let mut f = File::create("foo.txt").unwrap();
+/// f.write_all(b"Hello, world!").unwrap();
+///
+/// let mut f = File::open("foo.txt").unwrap();
+/// let mut s = String::new();
+/// f.read_to_string(&mut s).unwrap();
+/// assert_eq!(s, "Hello, world!");
+/// ```
+pub struct File {
+    inner: fs_imp::File,
+    path: PathBuf,
+}
+
+/// Metadata information about a file.
+///
+/// This structure is returned from the `file_attr` function or method and
+/// represents known metadata about a file such as its permissions, size,
+/// modification times, etc.
+pub struct FileAttr(fs_imp::FileAttr);
+
+/// Iterator over the entries in a directory.
+///
+/// This iterator is returned from the `read_dir` function of this module and
+/// will yield instances of `io::Result<DirEntry>`. Through a `DirEntry`
+/// information like the entry's path and possibly other metadata can be
+/// learned.
+pub struct ReadDir(fs_imp::ReadDir);
+
+/// Entries returned by the `ReadDir` entry.
+///
+/// An instance of `DirEntry` represents an entry inside of a directory on the
+/// filesystem. Each entry can be inspected via methods to learn about the full
+/// path or possibly other metadata through per-platform extension traits.
+pub struct DirEntry(fs_imp::DirEntry);
+
+/// Options and flags which can be passed to `File::open_opts`.
+///
+/// This builder exposes the ability to configure how a `File` is opened and
+/// what operations are permitted on it if it is successfully open.
+#[derive(Clone)]
+pub struct OpenOptions(fs_imp::OpenOptions);
+
+/// Representation of the various permissions on a file.
+///
+/// This module only currently provides one bit of information, `readonly`,
+/// which is exposed on all currently supported platforms. Unix-specific
+/// functionality, such as mode bits, is available through the
+/// `os::unix::FilePermissionsExt` trait.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FilePermissions(fs_imp::FilePermissions);
+
+impl File {
+    /// Open a file at `path` with the options specified by `opts`.
+    ///
+    /// # Error
+    ///
+    /// This function will return an error under a number of different
+    /// circumstances, to include but not limited to:
+    ///
+    /// * Opening a file that does not exist with read access.
+    /// * Attempting to open a file with access that the user lacks
+    ///   permissions for
+    /// * Filesystem-level errors (full disk, etc)
+    pub fn open_opts<P: AsPath + ?Sized>(path: &P, opts: &OpenOptions)
+                                         -> io::Result<File> {
+        let path = path.as_path();
+        let inner = try!(fs_imp::File::open(path, &opts.0));
+
+        // On *BSD systems, we can open a directory as a file and read from
```

Comment:
We actually [explicitly enable that today](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/fs.rs#L206-L207) but I made sure to not carry it over to the current implementation